frontend/test/pytest/test_adjoint.py

```diff
@@ -0,0 +1,210 @@
+# Copyright 2022-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Unit tests for the Catalyst adjoint function.
+"""
+
+from functools import partial
+
+import jax.numpy as jnp
+import pennylane as qml
+import pennylane.numpy as pnp
+import pytest
+from numpy.testing import assert_allclose
+from pennylane import adjoint as PL_adjoint
+
+from catalyst import adjoint as C_adjoint
+from catalyst import qjit
+
+# pylint: disable=missing-function-docstring
+
```

Comment:
@grwlf while the below are good functional/integration tests, I strongly recommend adding unit tests. That is, tests that explicitly check whether `adjoint` is correctly generating the circuit structure you expect.

This could be done in a multitude of ways; one approach could be to use pytest fixtures to inspect internal datastructures.

Reply:
@josh146 please check the [MLIR unit-tests](https://github.com/PennyLaneAI/catalyst/blob/dce83f9bcce0aedef254c5b2ea89af45146eff35/mlir/test/Quantum/AdjointTest.mlir#L18) ? I believe here I explicitly check the output structure.

Reply:
Thanks @grwlf! I would defer to @dime10 here - happy with his decision :)

Reply:
Yes for compiler transforms we have MLIR unit tests